src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorHoverProvider.ts

```diff
@@ -0,0 +1,79 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { RazorLanguageFeatureBase } from './RazorLanguageFeatureBase';
+
+export class RazorHoverProvider
+    extends RazorLanguageFeatureBase
+    implements vscode.HoverProvider {
+
+    public async provideHover(
+        document: vscode.TextDocument, position: vscode.Position,
+        token: vscode.CancellationToken) {
+
+        const projection = await this.getProjection(document, position, token);
+        if (!projection) {
+            return;
+        }
+
+        const results = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            projection.uri,
+            projection.position);
+
+        if (!results || results.length === 0) {
+            return;
+        }
+
+        // At the vscode.HoverProvider layer we can only return a single hover result. Because of this limitation we need to
+        // be smart about combining multiple hovers content or only take a single hover result. For now we'll only take one
+        // of them and then based on user feedback we can change this approach in the future.
+        const applicableHover = results.filter(item => item.range)[0];
+        if (!applicableHover) {
+            // No hovers available with a range.
+            return;
+        }
+
+        // Re-map the projected hover range to the host document range
+        const remappedResponse = await this.serviceClient.mapToDocumentRange(
+            projection.languageKind,
+            applicableHover.range!,
+            document.uri);
+
+        if (!remappedResponse) {
+            // Couldn't remap the projected hover location, there's no hover information available.
+            return;
+        }
+
+        if (document.version !== remappedResponse.hostDocumentVersion) {
+            // This hover result is for a different version of the text document, bail.
+            return;
+        }
+
+        const rewrittenContent = new Array<vscode.MarkedString>();
+        for (const content of applicableHover.contents) {
+            // For some reason VSCode doesn't respect the hover contents as-is. Because of this we need to look at each permutation
```

Comment:
@jrieken I found this super interesting. I doubt its intended though.

Reply:
That sounds interesting... What type is `applicableHover.contents`? Is that from a different library or is this a vscode-type? This might be an instanceof-check that's fooling us

Reply:
When debugging this, should end up here: https://github.com/microsoft/vscode/blob/9727eb60500657cd2c30e4a3992e17940c2cefba/src/vs/workbench/api/common/extHostTypeConverters.ts#L243 and I am curious to know what's failing there

Reply:
It looks like a MarkedString (JSON.stringify result):
```JSON
[
 {
  "_value": "```csharp\nViewDataDictionary<dynamic> Microsoft.AspNetCore.Mvc.Razor.RazorPage<dynamic>.ViewData\n```\n",
  "_isTrusted": false
 },
 {
  "_value": "",
  "_isTrusted": false
 }
]
```

with each entry being a MarkdownContent:
![image](https://user-images.githubusercontent.com/2008729/69190367-1da6bd80-0ad5-11ea-9b74-024eb716654d.png)


If you want to investigate it further you can get into a similar state by doing:
1. Run `./build.cmd -test` script at root of repo
2. Open root of repo in VSCode
3. Run the `Run Extension` VSCode command
4. Wait for O# log to stop spewing (let it initialize)
5. Open `MvcComponents/Views/Home/Index.cshtml`
6. Put a breakpoint in RazorHoverProvider.ts
7. Hover over `    ViewData["Title"] = "Home Page";`

Or I'm also down to schedule a meeting so we can pair debug together, whatever you think works better.

Reply:
Also I debugged to the bit of code you mentioned and I did see something a little funky. When OmniSharp returns their hover result this is what it initially looks like:
![image](https://user-images.githubusercontent.com/2008729/69192256-1681ae80-0ad9-11ea-871d-969c6c8bc7d1.png)

It seems like you guys are converting it into the `applicableHover` bits above at some point which can't be fed back to VSCode for presentation.